include/envoy/http/original_ip_detection.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/address.h"
+
+namespace Envoy {
+namespace Http {
+
+struct OriginalIPDetectionParams {
+  Http::RequestHeaderMap& request_headers;
+};
+
+/**
+ * Interface class for original IP detection extensions.
+ */
+class OriginalIPDetection {
+public:
+  virtual ~OriginalIPDetection() = default;
+
+  /**
+   * Detect the final remote address.
+   *
+   * @param param supplies the OriginalIPDetectionParams params for detection.
+   */
+  virtual Network::Address::InstanceConstSharedPtr
```

Comment:
This API seems to allow discovery of the original client IP, but not the original load-balancer/server ip:port.  Do we also need that extension point?

Also, consider changing the method name to detectRemoteIp to match function comments.

Reply:
wouldn't this extension point be enough for port discovery too (given an Instance may contain port too)? Naming and comments could use some more work tho.